packages/vm/src/context.rs

```diff
@@ -212,29 +209,26 @@ mod iter_support {
             Ok(o) => o,
             Err(_) => return ERROR_SCAN_INVALID_ORDER,
         };
-        let storage = take_storage::<S, Q>(ctx);
-        if let Some(store) = storage {
+        let mut res = ERROR_NO_STORAGE;
+        with_storage_from_context::<S, Q, _>(ctx, |store| {
             let iter = store.range(start.as_deref(), end.as_deref(), order);
             // Unsafe: I know the iterator will be deallocated before the storage as I control the lifetime below
             // But there is no way for the compiler to know. So... let's just lie to the compiler a little bit.
             let live_forever: Box<dyn Iterator<Item = KV> + 'static> =
                 unsafe { mem::transmute(iter) };
-            leave_iterator::<S, Q>(ctx, live_forever);
-            leave_storage::<S, Q>(ctx, Some(store));
-            SUCCESS
-        } else {
-            ERROR_NO_STORAGE
-        }
+            set_iterator::<S, Q>(ctx, live_forever);
+            res = SUCCESS;
+        });
+        res
     }
 
     pub fn do_next<S: Storage, Q: Querier>(ctx: &Ctx, key_ptr: u32, value_ptr: u32) -> i32 {
-        let mut iter = match take_iterator::<S, Q>(ctx) {
-            Some(i) => i,
-            None => return ERROR_NEXT_INVALID_ITERATOR,
-        };
-        // get next item and return iterator
-        let item = iter.next();
-        leave_iterator::<S, Q>(ctx, iter);
+        let mut item: Option<KV> = None;
+        with_iterator_from_context::<S, Q, _>(ctx, |iter| {
+            item = iter.next();
+        });
+        // TODO: handle this as well
+        // None => return ERROR_NEXT_INVALID_ITERATOR,
```

Comment:
Could `with_iterator_from_context` return a `Result<()>` to handle the case that the callback was not executed?

Reply:
I added the Result type to Storage, which also allows passing data out of the closure easier.

@webmaster128 
Please look at https://github.com/CosmWasm/cosmwasm/pull/234/commits/721754744ccf9111e79b8764d205c7ce760725e9 and let me know if you like the approach, and I will update the other ones

Reply:
Yes, nice. I was about to use the same model some days ago in a draft for #216

Reply:
I will try to update it now.